core/src/main/scala/org/ensime/indexer/DatabaseService.scala

```diff
@@ -28,83 +32,76 @@ class DatabaseService(dir: File) extends SLF4JLogging {
     (ds, Database.forDataSource(ds))
   }
 
-  def shutdown(): Unit = {
+  def shutdown()(implicit ec: ExecutionContext): Future[Unit] = for {
     // call directly - using slick withSession barfs as it runs a how many rows were updated
     // after shutdown is executed.
-    db.createSession().createStatement() execute "shutdown;"
-    datasource.close()
-  }
+    _ <- db.run(sqlu"shutdown")
+    _ <- db.shutdown
+    _ = datasource.close()
+  } yield ()
 
   if (!dir.exists) {
     log.info("creating the search database")
     dir.mkdirs()
-    db withSession { implicit s =>
-      (fileChecks.ddl ++ fqnSymbols.ddl).create
-    }
+    Await.result(
+      db.run(
+        (fileChecks.schema ++ fqnSymbols.schema).create
+      ),
+      DB_CREATE_TIMEOUT
+    )
   }
 
   // TODO hierarchy
   // TODO reverse lookup table
 
   // file with last modified time
-  def knownFiles(): List[FileCheck] = db.withSession { implicit s =>
-    (for (row <- fileChecks) yield row).list
-  }
+  def knownFiles(): Future[List[FileCheck]] = db.run(fileChecks.to[List].result)
+
+  def removeFiles(files: List[FileObject])(implicit ec: ExecutionContext): Future[Int] =
+    db.run {
+      val restrict = files.map(_.getName.getURI)
 
-  def removeFiles(files: List[FileObject]): Int = db.withSession { implicit s =>
-    val restrict = files.map(_.getName.getURI)
-    val q1 = for {
-      row <- fqnSymbols
-      if row.file inSet restrict
-    } yield row
-    q1.delete
-
-    val q2 = for {
-      row <- fileChecks
-      if row.filename inSet restrict
-    } yield row
-    q2.delete
+      (
+        fqnSymbols.filter(_.file inSet restrict).delete
+        andThen
+        fileChecks.filter(_.filename inSet restrict).delete
+      )
+    }
+
+  private val timestampsQuery = Compiled {
+    filename: Rep[String] => fileChecks.filter(_.filename === filename).map(_.timestamp).take(1)
   }
 
-  def outOfDate(f: FileObject): Boolean = db withSession { implicit s =>
+  def outOfDate(f: FileObject)(implicit ec: ExecutionContext): Future[Boolean] = {
     val uri = f.getName.getURI
     val modified = f.getContent.getLastModifiedTime
 
-    val query = for {
-      filename <- Parameters[String]
-      u <- fileChecks if u.filename === filename
-    } yield u.timestamp
-
-    query(uri).list.map(_.getTime).headOption match {
-      case Some(timestamp) if timestamp < modified => true
-      case Some(_) => false
-      case _ => true
-    }
+    db.run(
+      timestampsQuery(uri).result.headOption.map(_.map(_.getTime < modified).getOrElse(true))
```

Comment:
You probably know that: But the nested maps are because of the added Future Monad, now Future[Option[...]]. I find nested Monads are generally confusing. For the handling of the inner one here, I don't find the match helping my understanding. Happy to bring it back if you want. Just tell me what's needed to get this merged.